### ISSUE TITLE
Lost input headers when using https-client with self-signed certificates

### ISSUE THREAD

**Issue by nifigase** [libevent/libevent#233]
I try to use https-client with servers having self-signed certificates.
Connection is established successfully (with -ignore-cert), but when evhttp is about to read HTTP headers, the actual input starts with response body, so headers parsing procedure fails.
I compiled libevent with CFLAGS="-O0 -fno-inline -DUSE_DEBUG"
The following debug output is for one of public https servers with self-signed certificate.
```
$ bin/https-client -ignore-cert -url https://www.pcwebshop.co.uk:443 
[debug] Attempting connection to www.pcwebshop.co.uk:443

[debug] evhttp_add_header: key: Host val: www.pcwebshop.co.uk

[debug] evhttp_add_header: key: Connection val: close

[debug] event_del: 0xb69e10 (fd -1), callback 0x7f55f397dfca
[debug] event_del: 0xb69e90 (fd -1), callback 0x7f55f397dfca
[debug] event_add: event: 0xb69e10 (fd 6), EV_READ    call 0x7f55f397dfca
[debug] Epoll ADD(1) on fd 6 okay. [old events were 0; read change was 1; write change was 0; close change was 0]
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dfca
[debug] Epoll MOD(5) on fd 6 okay. [old events were 2; read change was 0; write change was 1; close change was 0]
[debug] event_add: event 0xb69e90, timeout in 45 seconds 0 useconds, call 0x7f55f397dfca
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dfca
[debug] event_add: event 0xb69e90, timeout in 45 seconds 0 useconds, call 0x7f55f397dfca
[debug] timeout_next: event: 0xb69e90, in 44 seconds, 999983 useconds
[debug] epoll_dispatch: epoll_wait reports 1
[debug] event_active: 0xb69e90 (fd 6), res 4, callback 0x7f55f397dfca
[debug] event_process_active: event: 0xb69e90,  EV_WRITE  call 0x7f55f397dfca
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dfca
[debug] event_add: event 0xb69e90, timeout in 1427988416 seconds 365089 useconds, call 0x7f55f397dfca
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dfca
[debug] Epoll MOD(1) on fd 6 okay. [old events were 6; read change was 0; write change was 2; close change was 0]
[debug] event_add: event: 0xb69e10 (fd 6), EV_READ    call 0x7f55f397dfca
[debug] epoll_dispatch: epoll_wait reports 1
[debug] event_active: 0xb69e10 (fd 6), res 2, callback 0x7f55f397dfca
[debug] event_process_active: event: 0xb69e10, EV_READ   call 0x7f55f397dfca
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dfca
[debug] event_add: event: 0xb69e10 (fd 6), EV_READ    call 0x7f55f397dfca
[debug] epoll_dispatch: epoll_wait reports 1
[debug] event_active: 0xb69e10 (fd 6), res 2, callback 0x7f55f397dfca
[debug] event_process_active: event: 0xb69e10, EV_READ   call 0x7f55f397dfca
[debug] event_del: 0xb69e10 (fd 6), callback 0x7f55f397dfca
[debug] Epoll DEL(1) on fd 6 okay. [old events were 2; read change was 2; write change was 0; close change was 0]
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dfca
[debug] event_add: event: 0xb69e10 (fd 6), EV_READ    call 0x7f55f397db56
[debug] Epoll ADD(1) on fd 6 okay. [old events were 0; read change was 1; write change was 0; close change was 0]
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] Epoll MOD(5) on fd 6 okay. [old events were 2; read change was 0; write change was 1; close change was 0]
[debug] event_add: event 0xb69e90, timeout in 45 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_process_active: event_callback 0xb69fa0, closure 3, call 0x7f55f395d32e
[debug] evhttp_connection_cb: connected to "www.pcwebshop.co.uk:443" on 6

[debug] event_add: event: 0xb69e10 (fd 6), EV_READ   EV_TIMEOUT call 0x7f55f397db56
[debug] event_add: event 0xb69e10, timeout in 50 seconds 0 useconds, call 0x7f55f397db56
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] event_add: event 0xb69e90, timeout in 50 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_del: 0xb69e10 (fd 6), callback 0x7f55f397db56
[debug] Epoll MOD(4) on fd 6 okay. [old events were 6; read change was 2; write change was 0; close change was 0]
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] event_add: event 0xb69e90, timeout in 50 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
[debug] Epoll DEL(4) on fd 6 okay. [old events were 4; read change was 0; write change was 2; close change was 0]
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] Epoll ADD(4) on fd 6 okay. [old events were 0; read change was 0; write change was 1; close change was 0]
[debug] event_add: event 0xb69e90, timeout in 50 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
[debug] Epoll DEL(4) on fd 6 okay. [old events were 4; read change was 0; write change was 2; close change was 0]
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] Epoll ADD(4) on fd 6 okay. [old events were 0; read change was 0; write change was 1; close change was 0]
[debug] event_add: event 0xb69e90, timeout in 50 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
[debug] Epoll DEL(4) on fd 6 okay. [old events were 4; read change was 0; write change was 2; close change was 0]
[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] Epoll ADD(4) on fd 6 okay. [old events were 0; read change was 0; write change was 1; close change was 0]
[debug] event_add: event 0xb69e90, timeout in 50 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
[debug] Epoll DEL(4) on fd 6 okay. [old events were 4; read change was 0; write change was 2; close change was 0]
[debug] evhttp_write_buffer: preparing to write buffer

[debug] event_add: event: 0xb69e90 (fd 6),  EV_WRITE  EV_TIMEOUT call 0x7f55f397dbb8
[debug] Epoll ADD(4) on fd 6 okay. [old events were 0; read change was 0; write change was 1; close change was 0]
[debug] event_add: event 0xb69e90, timeout in 50 seconds 0 useconds, call 0x7f55f397dbb8
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
[debug] Epoll DEL(4) on fd 6 okay. [old events were 4; read change was 0; write change was 2; close change was 0]
[debug] event_add: event: 0xb69e10 (fd 6), EV_READ   EV_TIMEOUT call 0x7f55f397db56
[debug] Epoll ADD(1) on fd 6 okay. [old events were 0; read change was 1; write change was 0; close change was 0]
[debug] event_add: event 0xb69e10, timeout in 50 seconds 0 useconds, call 0x7f55f397db56
[debug] event_process_active: event_callback 0xb69fa0, closure 3, call 0x7f55f395d32e
[debug] timeout_next: event: 0xb69e10, in 50 seconds, 0 useconds
[debug] epoll_dispatch: epoll_wait reports 1
[debug] event_active: 0xb69e10 (fd 6), res 2, callback 0x7f55f397db56
[debug] event_process_active: event: 0xb69e10, EV_READ   call 0x7f55f397db56
[debug] event_add: event: 0xb69e10 (fd 6), EV_READ   EV_TIMEOUT call 0x7f55f397db56
[debug] event_add: event 0xb69e10, timeout in 1427988421 seconds 565523 useconds, call 0x7f55f397db56
[debug] event_process_active: event_callback 0xb69fa0, closure 3, call 0x7f55f395d32e
[debug] evhttp_parse_http_version: bad version <!DOCTYPE on message 0xb6a1d0 from (null)
[debug] evhttp_read_firstline: bad header lines on 6

[debug] event_del: 0xb69e10 (fd 6), callback 0x7f55f397db56
[debug] Epoll DEL(1) on fd 6 okay. [old events were 2; read change was 2; write change was 0; close change was 0]
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
[debug] event_del: 0xb69e10 (fd 6), callback 0x7f55f397db56
[debug] event_del: 0xb69e90 (fd 6), callback 0x7f55f397dbb8
some request failed - no idea which one though!
socket error = Success (0)
[debug] event_base_loop: no events registered.
[debug] event_callback_finalize_many_: 3 events finalizing
[debug] event_del: 0xb69e10 (fd -1), callback 0x7f55f397db56
[debug] event_del: 0xb69e90 (fd -1), callback 0x7f55f397dbb8
[debug] event_del: 0xb69e10 (fd -1), callback 0x7f55f397db56
[debug] event_del: 0xb69e10 (fd -1), callback 0x7f55f395edf4
[debug] event_base_free_: 1 events were still set in base
```
See the line "[debug] evhttp_parse_http_version: bad version <!DOCTYPE on message 0xb6a1d0 from (null)"
For other sites with self-signed certificates the result is the same.
For sites with valid certificates everything works even with -ignore-cert option.

**Comment by nifigase** [comment]
I finally figured out what's going on. The reason of strange behavior is a bug in https-client.c - on line 278:
```
-       if (path == NULL) {
+       if (path == NULL || strlen(path) == 0) {
```
For URLs without path like "https://host:port" the path is not null, but empty string. So path is not set to anything and HTTP server sees incorrect request with no path ("GET  HTTP/1.1"), issues 400 error, and the reaction of libevent client code needs further investigation..
But now I 'm going to make a pull request with the fix in client.

**Comment by azat** [comment]
See #234

**Comment by ellzey** [comment]
Will be reviewing this and the suggested patch tonight. Sorry for the delay.

**Comment by azat** [comment]
Fixed by #234
